src/include/daos_pipeline.h

```diff
@@ -0,0 +1,256 @@
+ *
+ * \param[in,out]	pipeline	Pipeline object.
+ *
+ * \param[in]		node		Node object to be added to the pipeline.
+*/
+int
+daos_pipeline_push(daos_pipeline_t *pipeline, daos_pipeline_node_t *node);
+
+/**
+ * Adds a new filter object to the pipeline node object 'node'.
+ *
+ * \param[in,out]	node	Pipeline node object.
+ *
+ * \param[in]		filter	Filter object to be added to a pipeline node.
+ */
+int
+daos_pipeline_node_push(daos_node_t *node, daos_pipeline_filter_t *filter);
+
+/**
+ * Checks that a pipeline object is well built. If the pipeline object is well
+ * built, the function will return 0 (no error).
+ *
+ * \param[in]		pipeline	Pipeline object.
+ */
+int
+daos_pipeline_check(daos_pipeline_t *pipeline);
+
+/**
+ * Runs a pipeline on DAOS, returning objects and/or aggregated results.
+ *
+ * \param[in]		pipeline	Pipeline object.
+ *
+ * \param[in]		th		Optional transaction handle. Use
+ *					DAOS_TX_NONE for an independent
+ *					transaction.
+ *
+ * \param[in]		flags		Conditional operations.
+ *
+ * \param[in]		dkey		Optional dkey. When passed, no iteration
+ *					is done and processing isonly performed
+ *					on this specific dkey.
+ *
+ * \param[in]		nr_iods		Number of I/O descriptors in the iods
+ *					table.
+ *
+ * \param[in,out]	iods		[in]: Array of I/O descriptors. Each
+ *					descriptor is associated with a given
+ *					akey and describes the list of
+ *					record extents to fetch from the array.
+ *					[out]: Checksum of each extent is
+ *					returned via
+ *					\a iods[]::iod_csums[]. If the record
+ *					size of an extent is unknown (i.e. set
+ *					to DAOS_REC_ANY as input), then the
+ *					actual record size will be returned in
+ *					\a iods[]::iod_size.
+ *
+ * \param[in,out]	anchor		Hash anchor for the next call, it should
+ *					be set to zeroes for the first call, it
+ *					should not be changed by caller
+ *					between calls.
+ *
+ * \param[in,out]	nr_kds		[in]: Number of key descriptors in
+ *					\a kds.
+ *					[out:] Number of returned key descriptors.
+ *
+ * \param[in,out]	kds		[in]: Optional preallocated array of \nr
+ *					key descriptors.
+ *					[out]: Size of each individual key along
+ *					with checksum type and size stored just
+ *					after the key in \a sgl_keys.
+ *
+ * \param[out]		sgl_keys	Optional sgl storing all dkeys to be
+ *					returned.
+ *
+ * \param[out]		sgl_recx	Optional sgl storing all the records to
+ *					be returned.
+ *
+ * \param[out]		sgl_agg		Optional sgl with the returned value of
+ *					the aggregator(s).
+ *
+ * \param[in]		ev		Completion event. It is optional.
+ *					Function will run in blocking mode if
+ *					\a ev is NULL.
+ */
+int
+daos_pipeline_run(daos_pipeline_t *pipeline, daos_handle_t th, uint64_t flags,
+		  daos_key_t *dkey, uint32_t nr_iods, daos_iod_t *iods,
+		  daos_anchor_t *anchor, uint32_t *nr_kds, daos_key_desc_t *kds,
+		  d_sg_list_t *sgl_keys, d_sg_list_t *sgl_recx,
+		  d_sg_list_t *sgl_agg, daos_event_t *ev);
+
```

Comment:
Hmm, you put oh in daos_pipeline_t, but leave dkey, iods, iods_nr in the API interface?  why? 

Probably you should put  oh/dkey/akey/iods  in both daos_pipeline_t and daos_pipeline_node_t,  then you can filter them optional in each level?

Reply:
_**Hmm, you put oh in daos_pipeline_t, but leave dkey, iods, iods_nr in the API interface? why?**_
No particular reason. Those can be moved to the daos_pipeline_t.

_**Probably you should put oh/dkey/akey/iods in both daos_pipeline_t and daos_pipeline_node_t, then you can filter them optional in each level?**_
I am not sure I understand your suggestions here. nodes are part of a pipeline, and pipelines only affect a DAOS object. It is enough to include that on the high level daos_pipeline_t object. We don't want to support overly complex pipelines at this stage (like those running over multiple DAOS objects). Maybe that can be a future extension, when we have this up and running.

Reply:
then you can only filter a single dkey per object(pipeline call)?  
And also iods is in daos_pipeline_t, but akey is in daos_pipeline_node_t,  so it means it can only filter  single dkey + iods  for different akey?  is this what you want? or I miss sth?

Note: there is also akey in daos_iod_t (iod->iod_name) as well.

Reply:
_**then you can only filter a single dkey per object(pipeline call)?**_
No, that is a special case, suggested by @johannlombardi . The general use case is to filter a whole object. You can filter (return a subset of records) by value of dkeys and/or akeys. You can also return an aggregation of all records or a subset (output from previous filter step).

And also iods is in daos_pipeline_t, but akey is in daos_pipeline_node_t, so it means it can only filter single dkey + iods for different akey? is this what you want? or I miss sth?

iods in daos_pipeline_t just indicate a projection. Meaning, of all records returned (the ones passed filters), you only return a subset of akeys. The akey in daos_pipeline_filter_t is to indicate which akey is used in the filter. For example, filter by age<5, we indicate that the akey of the filter is 'age'. But then, in iods, we may want to return 'age' and 'name' from the records.

Reply:
.>You can filter (return a subset of records) by value of dkeys and/or akeys.

I might miss sth, how do you specify multiple dkeys with single object (one daos_pipeline_run).

>iods in daos_pipeline_t just indicate a projection. Meaning, of all records returned (the ones passed filters), you only return a subset of akeys. The akey in daos_pipeline_filter_t is to indicate which akey is used in the filter. For example, filter by age<5, we indicate that the akey of the filter is 'age'. But then, in iods, we may want to return 'age' and 'name' from the records.

Ah, thanks for clarification.

Reply:
_**_I might miss sth, how do you specify multiple dkeys with single object (one daos_pipeline_run)._**_

You don't, that is why there is no 'dkey' variable in daos_pipeline_filter_t. What you say is to filter by **the** dkey. So you can filter by dkey<5. That is 3 filter objects. One for the function "<" and 2 parameters. In the case of the dkey parameter, the filter object is just of type DAOS_FILTER_DKEY, but we don't need to specify which dkey.

Reply:
ah, I see. thanks for clarification.